### PR TITLE
Ensure we only iterate over process groups

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

Fixes: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/918 (Ensure in reconcilers that we iterate over ProcessGroups not Instances/Pods )

## Type of change

*Please select one of the options below.*

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

# Discussion

- 

# Testing

Unit + local 

# Documentation

Adjusted.

# Follow-up

-

### CODE REVIEW

File: controllers/replace_misconfigured_process_groups.go [hunk 1]
```diff
@@ -34,67 +34,69 @@ import (
...
+			needsPVCRemoval, err := instanceNeedsRemovalForPVC(cluster, pvc)
 			if err != nil {
 				return &Requeue{Error: err}
 			}
 
-			if len(instances) > 0 {
-				processGroupStatus.Remove = true
-				hasNewRemovals = true
+			if needsPVCRemoval {
+				if hasPod {
```
Comment (johscheuer) [inline controllers/replace_misconfigured_process_groups.go:76]:
```suggestion
			if needsPVCRemoval && hasPod {
```

I'm actually not sure if we really need to check the `hasPod` condition. I rather think we can drop it.

File: controllers/delete_pods_for_buggification.go [hunk 2]
```diff
@@ -61,10 +55,17 @@ func (d DeletePodsForBuggification) Reconcile(r *FoundationDBClusterReconciler,
 		}
 	}
 
-	for _, pod := range pods {
-		instanceID := GetProcessGroupID(cluster, pod)
-		_, pendingRemoval := removals[instanceID]
-		if pendingRemoval {
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		if processGroup.Remove {
+			logger.V(1).Info("Ignore removed Pod",
```
Comment (johscheuer) [inline controllers/delete_pods_for_buggification.go:60]:
```suggestion
			logger.V(1).Info("Ignore process group marked for removal",
```

File: controllers/update_labels.go [hunk 3]
```diff
@@ -37,60 +35,58 @@ type UpdateLabels struct{}
...
-		processClass := internal.GetProcessClassFromMeta(cluster, pvc.ObjectMeta)
-		instanceID := internal.GetProcessGroupIDFromMeta(cluster, pvc.ObjectMeta)
+	pvcMap := internal.CreatePVCMap(cluster, pvcs)
 
-		metadata := internal.GetPvcMetadata(cluster, processClass, instanceID)
-		if metadata.Annotations == nil {
-			metadata.Annotations = make(map[string]string, 1)
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		if processGroup.Remove {
+			logger.V(1).Info("Ignore removed Pod",
```
Comment (johscheuer) [inline controllers/update_labels.go:54]:
```suggestion
			logger.V(1).Info("Ignore process group marked for removal",
```

File: controllers/update_sidecar_versions.go [hunk 4]
```diff
@@ -44,7 +44,16 @@ func (u UpdateSidecarVersions) Reconcile(r *FoundationDBClusterReconciler, conte
 		return &Requeue{Error: err}
 	}
 	upgraded := false
-	for _, pod := range pods {
+
+	podMap := internal.CreatePodMap(cluster, pods)
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		pod, ok := podMap[processGroup.ProcessGroupID]
```
Comment (johscheuer) [inline controllers/update_sidecar_versions.go:50]: TODO: Check if the PG is removed and if so skip it.

File: controllers/update_pod_config.go [hunk 5]
```diff
@@ -69,7 +60,7 @@ func (u UpdatePodConfig) Reconcile(r *FoundationDBClusterReconciler, context ctx
 			continue
 		}
 
-		pod, ok := podProcessGroupMap[processGroup.ProcessGroupID]
+		pod, ok := podMap[processGroup.ProcessGroupID]
```
Comment (johscheuer) [inline controllers/update_pod_config.go:63]: We should also skip process groups marked for removal here.

**Review by brownleej — APPROVED** [review]
👍

**Pull request merged**